shared-libs/infodoc/src/infodoc.js

```diff
@@ -0,0 +1,255 @@
+            })
+            .catch(err => {
+              if (err.status !== 409) {
+                throw err;
+              }
+
+              // conflict, try the whole thing again from the top
+              return getInfoDoc(change);
+            });
+        });
+    });
+};
+
+const deleteInfoDoc = change => {
+  return db.sentinel
+    .get(getInfoDocId(change.id))
+    .then(doc => {
+      doc._deleted = true;
+      return db.sentinel.put(doc);
+    })
+    .catch(err => {
+      if (err.status !== 404) {
+        throw err;
+      }
+    });
+};
+
+const updateTransition = (change, transition, ok) => {
+  const info = change.info;
+  info.transitions = info.transitions || {};
+  info.transitions[transition] = {
+    last_rev: change.doc._rev,
+    seq: change.seq,
+    ok: ok,
+  };
+};
+
+const saveTransitions = change => {
+  return db.sentinel.get(getInfoDocId(change.id))
+    .then(doc => {
+      doc.transitions = change.info.transitions || {};
+      return db.sentinel.put(doc);
+    })
+    .catch(err => {
+      if (err.status !== 409) {
+        throw err;
+      }
+
+      return saveTransitions(change);
+    });
+};
+
+const bulkGet = changes => {
+  const infoDocs = [];
+
+  if (!changes || !changes.length) {
+    return Promise.resolve();
+  }
+
+  return findInfoDocs(db.sentinel, changes)
+    .then(result => {
+      const missing = [];
+      result.forEach(row => {
+        if (!row.doc) {
+          missing.push({ id: getDocId(row.key) });
+        } else {
+          infoDocs.push(row.doc);
+        }
+      });
+
+      if (!missing.length) {
+        return [];
+      }
+
+      return findInfoDocs(db.medic, missing);
+    })
+    .then(result => {
+      result.forEach(row => {
+        if (!row.doc) {
+          infoDocs.push(blankInfoDoc(getDocId(row.key)));
+        } else {
+          row.doc.legacy = true;
+          infoDocs.push(row.doc);
+        }
+      });
+
+      return infoDocs;
+    });
+};
+
+const bulkUpdate = infoDocs => {
+  const legacyDocs = [];
+
+  if (!infoDocs || !infoDocs.length) {
+    return Promise.resolve();
+  }
+
+  infoDocs.forEach(doc => {
+    if (doc.legacy) {
+      delete doc.legacy;
+      legacyDocs.push(Object.assign({ _deleted: true }, doc));
+      delete doc._rev;
+    }
+  });
+
+  return db.sentinel.bulkDocs(infoDocs).then(results => {
+    if (legacyDocs.length) {
+      // Delete intentionally out of control flow for performance
+      db.medic.bulkDocs(legacyDocs);
+    }
+
+    const conflictingInfoDocs = infoDocs
+      .filter((_, idx) => results[idx].error === 'conflict');
+
+    if (conflictingInfoDocs.length > 0) {
+      // Attempt an intelligent merge based on responsibilities For right now this is only the
+      // transitions block. If the caller of this code (sentinel) needs to maintain other
+      // information in the infodoc or others call this code we would need to be smarter about how
+      // we merge.
+      return db.sentinel.allDocs({
```

Comment:
Interested in your thoughts here.

In the code as written API's write of the infodoc takes precedence with only sentinel's changes to the `transitions` parameter making it through on conflict.

I'm wondering if this should be reversed: sentinel taking precedence and API's version only being used for the date metadata.